### DISCUSSION TITLE
URS file not supported?

### DISCUSSION THREAD

**Discussion by architektpetr** [visose/Robots#107] [Q&A] [answered]
Hello. 

I just started using Robots instead of HAL3000. This plugin is well done and looks great. May I ask for help with four issues? (I use UR10): 

1. Robots generates .URS file which is not recognized by UR PolyScope. The file extension needs to be .script ... Is there a way how to change this automatically? Also, there is still the possibility that I do something wrong. 
2. The script won't run unless I call the function at the end (manual edit of .urs file). For example: 

def Program():
movej(...)
movej(...)
**Program()** _<<< this needs to be added at the end_

3. Also, my remote access to the robot doesn't work. But that might be just absence of my knowledges. Is that just that easy - plug robot in, set IP address and "play" the program? Or does it require some additional setup in PolyScope? 
4. Is it possible to edit acceleration? 

Thanks! 
Petr

**Comment by visose** [comment, marked as answer]
Hi, once you get the text output from the Create Program component you can use standard Grasshopper components to make those changes, like add Program() line to the end of save it with any file name and extension.

You must put the IP address of the robot in the Grasshopper component. There shouldn't be any additional step required in the UR pendant, except making sure the network settings is enabled and the robot is configured with a fixed IP.

The Speed parameter does have acceleration, but it's not exposed as an input. You can modify the value using a scripting component.

**Comment by architektpetr** [comment]
Thanks for comments! One last thing: Value list doesn't work automatically, but I have to input string from panel or so (for example "Batlett_UR10". The xml file is copied and unblocked as you described in the manual. 

Here is how I modified your component, so it exports .script file with proper syntax: 
(I was not able to upload .gh file, so it's [linked from my cloud](https://drive.google.com/file/d/1MOpv3jyYT4I10jvFAPrLT-AYHEfqwIOP/view?usp=sharing))

![image](https://user-images.githubusercontent.com/73644039/106495725-581f2f00-64bc-11eb-8d1c-850c0bbf0bad.png)

**Comment by visose** [comment]
You have to place a new Create RobotSystem component for the value list to populate with the list of robots.

**Comment by visose** [comment]
Hi, I will close the issue assuming this worked, you can reopen if not.

**Comment by architektpetr** [comment]
Hello

I tested the IP connection just today and it works well. Thanks for
suggestions.
There's still some trouble with the definition of the robot - the value
list doesn't work, I have to write "Bartlett-UR10" to a panel and plug it
in. But the plugin works fine. Just sometimes there are error messages
without any other hint of what is wrong.
Also, is there a way to set "time" input for PTP movement?

Best regards,

*Petr Vacek*
Architekt | Computational Designer | Maker

<https://www.linkedin.com/in/petr-vacek/>
<https://www.instagram.com/_petrvacek/>
<https://cz.pinterest.com/vacekpetr/boards/>
<https://www.facebook.com/petr.vacek.7739/>

petrvacek.com

+420 732 92 82 82

***@***.***

Prague, Czech Republic <https://goo.gl/maps/uYaR4pUhk92UXv2C9>


st 17. 2. 2021 v 11:48 odesílatel visose ***@***.***> napsal:

**Comment by robin-gdwl** [comment]
@architektpetr
The time input is not exposed in the standard grasshopper component same as the acceleration input. 
You need a script component and access the Speed-Parameter there and set a time or acceleration value. 
I recently did the same thing as documented here:
https://github.com/visose/Robots/issues/51#issuecomment-775253135 

Steps:
- add a C#-script component in grasshopper
- right click and select "manage assemblies" 
- add robots.dll file to the assembly list (it is located in the folder where you install gh-plugins)
- paste this code into the c# script 
```
var inputspeed = new Speed() ;
inputspeed.Time = time;

Robotspeed = inputspeed;
Print(inputspeed.Time.ToString());
```
- rename an input of the c#-script component to "time" 
- plug the script component into a "create target"-component

See the linked thread for more info.

**Comment by UnfoldAntwerp** [comment]
> 2\. The script won't run unless I call the function at the end (manual edit of .urs file). For example:
> 
> def Program(): movej(...) movej(...) **Program()** _<<< this needs to be added at the end_

I was running into the same issue. But unless I understand wrong your solution, it didn't work. I managed to get the script loaded by **removing** the `def Program():` line at the top of the script instead of adding anything to the end. Runs fine after that but you either need to select 'all files' when you want to load the script or change the extension to .script. I largely followed these instructions: https://robodk.com/doc/en/Robots-Universal-Robots.html#UR-Script for the rest.

**Reply by visose** [reply]
Both should work. The `def Program():` line declares a function. When sending it through the network, the `Program` function is automatically run, but it's not when it is added as script to a URP program. Removing the function declaration makes the content of the function top level code, which is always run. Calling the function after declaring it will also work.

I'll add to the todo list creating a proper URP and script files when using the `save program` component. And after maybe I could add uploading the program to the controller via sftp directly from Grasshopper.

**Reply by KonradJuenger** [reply]
In a different setting I used curl for uploading files generated with gh. -> It is not as clean, since it has to generate the file first and then upload but curl works very reliable and is preinstalled on mac and pc. I've modified + attached the file here: [curl.gh](https://github.com/KonradJuenger/gh_scripts/blob/main/curl.gh)

**Reply by visose** [reply]
From a quick search I did yesterday, I was thinking to try [this](https://github.com/sshnet/SSH.NET) library. It's cross-platform, the file can be a "stream" (don't have to save a temp file to disk before uploading), and don't have to start up an external process like with curl. 

The downside is relying on another dependency (another .dll). For a plugin this is less of a problem than with a single grasshopper file like your example.

**Reply by UnfoldAntwerp** [reply]
Happy to test if you've got something alpha level!

**Comment by visose** [comment]
Version 1.3.0 should now save a proper `.urp` program using the 'save program' component, and you can also opt to use the FTP server in the 'Remote connection' component.

The code is embedded inside the `.urp` file, so there is no need for a separate `.script` file.

If you input an IP address, it will use the secondary client interface to stream the program and run the play and pause commands. You can switch to the FTP back-end by specifying the IP in the format of `sftp://ip`. This will upload the program as a `.urp` file to the controller using the FTP server and load it. It uses the dashboard interface for loading, play, and pause. You can change the default values for username, password, or location of the programs folder, by using: `sftp://user:pass@ip/path/to/programs`

I have only tested this with the UR simulators (both CB and e-series), but it doesn't come with a sftp server. I installed one to test, but I might have gotten wrong the default path to the '/programs' folder.

**Reply by UnfoldAntwerp** [reply]
Super cool! Will give this a try once I have access to the UR10 again. Might not be for a while though.

**Reply by visose** [reply]
By the way, if anyone is using URSim, I'm running it on Windows under WSL2 and it works great, no need for VirtualBox or similar. You can use directly the distribution for Linux.

<img width="1234" alt="image" src="https://user-images.githubusercontent.com/5129614/169042333-1bdf1dd8-b54c-43bd-8264-d296e64187ec.png">

**Reply by robin-gdwl** [reply]
Thanks for the hint with running URsim on WSL2. Didnt even think of that.

**Reply by robin-gdwl** [reply]
The upload via FTP method is great. In my case I have to use the username "ur" instead of "root" the default settings did not work for me. I will also do some testing if it can alleviate problems with sending large programs.

**Reply by visose** [reply]
Great, @robin-gdwl can you confirm you didn't change the default folder path, and it is in fact `/programs`?

**Reply by visose** [reply]
> Note: In my testing with URsim and FTP connections the username set by default is not "root" but "ur".
I can not get FTP connections with root:easybot on a clean install of the latest version of URsim.
ur:easybot works.

The ftp server is not related to the UR software. You probably have a Linux instance with the main user named "ur" (a user of the operating system). There will also be an ftp server running that uses the users of the OS to log in (in this case, the "ur" user).
